src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs

```diff
@@ -431,28 +432,142 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
+                            case 1:
+                                switch (unmanagedCallingConventionTypes.Single().Name)
+                                {
+                                    case "CallConvCdecl":
+                                        actualCallKind = Cci.CallingConvention.CDecl;
+                                        actualUnmanagedCallingConvetionTypes = ImmutableHashSet<INamedTypeSymbolInternal>.Empty;
+                                        break;
+                                    case "CallConvStdcall":
+                                        actualCallKind = Cci.CallingConvention.Standard;
+                                        actualUnmanagedCallingConvetionTypes = ImmutableHashSet<INamedTypeSymbolInternal>.Empty;
+                                        break;
+                                    case "CallConvThiscall":
+                                        actualCallKind = Cci.CallingConvention.ThisCall;
+                                        actualUnmanagedCallingConvetionTypes = ImmutableHashSet<INamedTypeSymbolInternal>.Empty;
+                                        break;
+                                    case "CallConvFastcall":
+                                        actualCallKind = Cci.CallingConvention.FastCall;
+                                        actualUnmanagedCallingConvetionTypes = ImmutableHashSet<INamedTypeSymbolInternal>.Empty;
+                                        break;
+                                    default:
+                                        goto outerDefault;
+                                }
+                                break;
+
+                            default:
+outerDefault:
+                                actualCallKind = Cci.CallingConvention.Unmanaged;
+                                actualUnmanagedCallingConvetionTypes = unmanagedCallingConventionTypes;
+                                break;
+                        }
+                    }
+
+                    // The rules for matching a calling convention are:
+                    // 1. The CallKinds must match exactly
+                    // 2. If the CallKind is Unmanaged, then the set of calling convention types must match exactly, ignoring order
+                    //    and duplicates. We already have both sets in a HashSet, so we can just ensure they're the same length and
+                    //    that everything from one set is in the other set.
+
+                    if (!actualCallKind.IsCallingConvention(expectedConvention.CallKind))
+                    {
+                        results[i] = makeWrongCallingConvention(result);
+                        continue;
+                    }
+
+                    if (expectedConvention.CallKind.IsCallingConvention(Cci.CallingConvention.Unmanaged))
+                    {
+                        switch (expectedConvention.UnmanagedCallingConventionTypes, actualUnmanagedCallingConvetionTypes)
```

Comment:
>switch (expectedConvention.UnmanagedCallingConventionTypes, actualUnmanagedCallingConvetionTypes) [](start = 24, length = 97)

It looks like ConversionBase.HasImplicitPointerConversion simply uses SetEquals helper. Is there a reason to use custom implementation here? It feels like the calling convention comparison logic should be extracted into a helper and shared between ConversionsBase and OverloadResolution.

Reply:
I forgot about that method, thanks.

---
In reply to: [491117561](https://github.com/dotnet/roslyn/pull/47651#discussion_r491117561) [](ancestors = 491117561)

Reply:
Oh, actually, no, we can't use that method here. We have a set of `CustomModifier`s and a set of `INamedTypeSymbolInternal`, and I didn't feel that doing an allocation to get custom modifiers into a state that could be set equaled with a hashset was worth it.

---
In reply to: [491121777](https://github.com/dotnet/roslyn/pull/47651#discussion_r491121777) [](ancestors = 491121777,491117561)

Reply:
> We have a set of CustomModifiers and a set of INamedTypeSymbolInternal

I see.

---
In reply to: [491122866](https://github.com/dotnet/roslyn/pull/47651#discussion_r491122866) [](ancestors = 491122866,491121777,491117561)